---
title: "Circular Manhattan Plot, Violin Plot and Epistasis Heat Map"
output: github_document
---

<style type="text/css">
  
  
h1.title {/* Header 1 */
    font-size: 28px;
    font-family:  "Times New Roman", Times, serif;
    color: BLACK;
    #text-transform: Capitalize;
  }
  
h2 { /* Header 2 */
    font-size: 24px;
    font-family: "Times New Roman", Times, serif;
    color: Black;
    text-transform: none;
  }
h3 { /* Header 3 */
    font-size: 18px;
    font-family: "Times New Roman", Times, serif;
    color: DarkBlue;
    text-transform: none;
  }
  
h4 { /* Header 4 */
    font-size: 16px;
    font-family: "Times New Roman", Times, serif;
    color: Darkred;
    text-transform: none;
  }
  </style>

```{r knitr_init, echo=FALSE, cache=FALSE}
library(knitr)
library(rmdformats)

## Global options
options(max.print="75")
opts_chunk$set(echo=TRUE,
               cache=TRUE,
               prompt=FALSE,
               tidy=TRUE,
               comment=NA,
               message=FALSE,
               warning=FALSE,
               eval=FALSE)
opts_knit$set(width=75)
```

**Load the required libraries**
```{r}
      Packages <- c("dglm", "hglm", "gdata","readR","tidyr","ggplot",
                    "qqman", "CMplot",  "dplyr", "ggplot2", "SNPRelate",
                    "ggcorrplot", "statmod")
    lapply(Packages, library, character.only = TRUE)

```
## **Circular Manhattan Plot**

* Here we will draw the circular Manhattan plot using p-values including P.mean (equivalent to Traditional GWAS),  P.dispersion (obtained in DGLM analysis), and P.dispersion (obtained through HGLM analysis): Figure 1 in the manuscript.

### Load the GWAS, DGLM and HGLM outputs
**Description of the code:**

* Here we will load the data containing p.values for all the three including traditional GWAS, dglm and hglm analysis.
* We will create function to upload all the CSV files
* Combined the p-values from all the three files into one file and rename the columns
```{r}
# First we will create a function to upload all the 3 csv files function
      setwd("~/Box/Postdoc/Dataanalysis/Wheat_HWW_panel/data_manhattan_all")
      import.all<-function(mypath,mypattern,...)
      {
        tmp.list.1<-list.files(mypath, pattern=mypattern)
        tmp.list.2<-list(length=length(tmp.list.1))
        for (i in 1:length(tmp.list.1)){tmp.list.2[[i]]<-read.csv(tmp.list.1[i],...)}
        names(tmp.list.2)<-tmp.list.1
        tmp.list.2
      }
      
# Now upload the all the files
  csv.import<-import.all("~/Box/Postdoc/Dataanalysis/Wheat_HWW_panel/data_manhattan_all", "csv$",sep=",")
# here we define the separator of entries in the csv files to be comma.
      gwas_cd<-csv.import$gwas_cd.csv
      dglm_cd<-csv.import$dglm_cd.csv
      hglm_cd<-csv.import$hglm_cd.csv
  # subset the data
      colnames(gwas_cd)<-c("marker", "chrom", "pos", "gwas_p")
      dglm_cd<- dglm_cd[, c(1,5)]
      colnames(dglm_cd)<-c("marker","dglm_p.disp")
      hglm_cd<-hglm_cd[, c(1,4)]
      colnames(hglm_cd)<-c("marker", "hglm_p.disp")
# Now merge all the three data files
      all_p<-Reduce(merge, list(gwas_cd, dglm_cd, hglm_cd))
```

### Draw the circular Manhattan plot

```{r}
  CMplot(all_p,plot.type="c",r=0.4, col=matrix(c("grey30","slategray", NA,"lightskyblue1","lightseagreen", NA, "thistle","pink", NA),3,3,byrow=T),
         chr.labels=paste("",c("1A", "1B", "1D", "2A", "2B", "2D","3A", "3B", "3D","4A", "4B", "4D","5A", 
    "5B", "5D","6A", "6B", "6D", "7A", "7B", "7D", "UN"),sep=""), cir.legend=TRUE,cir.legend.cex=0.8,
         threshold=1e-5, outward=FALSE,amplify=TRUE,
         threshold.lty=c(1,2), signal.line=1,signal.col="orangered",threshold.col="blue",
         threshold.lwd=0.5, signal.cex=1.5,signal.pch=19, cir.chr.h=1.5,chr.den.col="black",file="jpg",memo="",dpi=300)
```
## **Violin plots**

* Here we are providing sample codes used ti draw the effective plots (Figure 2 in the manuscript) 
```{r}
ggplot(data_2a_final, aes(x=marker, y=avg1))+
  geom_violin(aes(fill = Genotype), trim = FALSE, position = position_dodge(0.9))+ 
  geom_hline(yintercept = 0.01309, color = "darkred", size =0.6,show.legend = TRUE,linetype = 2)+
  geom_boxplot(aes(fill = Genotype), width = 0.2, position = position_dodge(0.9))+
  scale_fill_manual(values = c("#00AFBB", "#E7B800"))+
  #stat_summary(fun.y=mean, shape=8, geom="point",size=2, aes(group=genotype), 
  #col="black") + #display mean as an asterisk
  #geom_text(data=chr_b_means, aes(label=cd, x=marker,y=cd), 
  #colour="black", size=3)
  labs(title = "", y="Cadmium conc. (mg/kg)", x="Markers")+
  theme_few() + #change background of the plot
  theme(plot.title = element_text(color="black", face="bold",size=12, hjust=0.5))+
  theme(axis.text.x=element_text(colour='black', size=12)) + #asthetics of x-axis text
  theme(axis.text.y=element_text(colour='black', size=12)) + #asthetics of y-axis text
  theme(axis.title.x = element_text(colour='black', size=12, vjust=0.0, face="bold")) + #asthetics of x-axis title
  theme(axis.title.y = element_text(colour='black', size=12, face="bold"))+
  theme(legend.title = element_text(colour="darkred", size=14, face="bold"),
        legend.text = element_text(colour="grey0", size=11, face="bold"))+
  guides(fill=guide_legend(title="Genotype"))
```

## **Heat map of epistasis and interaction effective plot**

* Here we are providing the sample codes used to draw the heat map of epistasis between markers associated with varaince-heterogeneity (Figure 3 of the manuscript) and interaction effective plot (Figure 4 of the manuscript)

```{r}
# Read the data file
epis<-read.csv(file="~/Box/Postdoc/Dataanalysis/Wheat_HWW_panel/Epistasis/
               Epistasis_output/epistasis_effectiveplot_DATA.csv", header=TRUE)
# Filter the two markers for interaction
epis_filter<-epis%>%filter(marker==c("IAAV3067", "IWA7579"))
# Create function for data summary
data_summary <- function(data, varname, groupnames){
  require(plyr)
  summary_func <- function(x, col){
    c(mean = mean(x[[col]], na.rm=TRUE),
      sd = sd(x[[col]], na.rm=TRUE))
  }
  data_sum<-ddply(data, groupnames, .fun=summary_func,
                  varname)
  data_sum <- rename(data_sum, c("mean" = varname))
  return(data_sum)
}
epis_sum <- data_summary(epis_filter, varname="phenotype", 
                    groupnames=c("marker", "allele"))
epis_sum$phenotype<-round(epis_sum$phenotype, digits=3)
epis_sum$sd<-round(epis_sum$sd, digits=3)
# Drop Na
epis_sum<-epis_sum%>%na.omit() 
# Draw the effetive plot
ggplot(epis_sum, aes(x=allele, y=phenotype, group=marker, color=marker)) +
  geom_errorbar(aes(ymin=phenotype-sd, ymax=phenotype+sd), width=.1) +
  geom_line() + geom_point()+
  scale_color_manual(values=c("darkred", "blue"))+
  labs(title="",x="", y = "Cadmium conc.(mg)")+
  theme_classic()+
  theme(axis.text= element_text(color = "black", size = 12))+
  theme (plot.title = element_text(),
         axis.title.y = element_text(color="black", size=12, face="bold")) +
  theme(legend.title = element_text(colour="black", size=14),
        legend.text = element_text(colour="black", size=12))+ # add and modify the legends
  guides(color=guide_legend(title=""))+
  theme(legend.position = c(0.15, 0.9))

# Now draw the HEAT MAP

epis_pvalues<-read.csv(file="~/Box/Postdoc/Dataanalysis/Wheat_HWW_panel/ Epistasis/Epistasis_output/epistasis_pvalues.csv",
                       header=TRUE, row.names = 1)
# Convert NA into zeros all the values into -log10
epis_pvalues[is.na(epis_pvalues)]<-1
epis_pvalues<-as.matrix(epis_pvalues)
epis_pvalues<- -log10(epis_pvalues)
# draw the corrplot
heatmap_epis<-ggcorrplot(epis_pvalues, method="square", show.diag = TRUE,
                         tl.cex = 6, show.legend = FALSE)
heatmap_epis
# draw the corrplot
heatmap_epis<-ggcorrplot(epis_pvalues, method="square", show.diag = TRUE,
                         tl.cex = 6, show.legend = FALSE)+
  theme(plot.margin = unit(c(0.5,0.5,2,2), "cm"))
  heatmap_epis
```
