---
title: "Principal component analysis (PCA)"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

 <style type="text/css">
  
h1.title {/* Header 1 */
    font-size: 28px;
    font-family:  "Times New Roman", Times, serif;
    color: BLACK;
    #text-transform: Capitalize;
  }
  
h2 { /* Header 2 */
    font-size: 24px;
    font-family: "Times New Roman", Times, serif;
    color: Black;
    text-transform: none;
  }
h3 { /* Header 3 */
    font-size: 18px;
    font-family: "Times New Roman", Times, serif;
    color: DarkBlue;
    text-transform: none;
  }
  
h4 { /* Header 4 */
    font-size: 16px;
    font-family: "Times New Roman", Times, serif;
    color: Darkred;
    text-transform: none;
  }
  </style>

```{r knitr_init, echo=FALSE, cache=FALSE}
library(knitr)
library(rmdformats)

## Global options
options(max.print="75")
opts_chunk$set(echo=TRUE,
               cache=TRUE,
               prompt=FALSE,
               tidy=TRUE,
               comment=NA,
               message=FALSE,
               warning=FALSE,
               eval=FALSE)
opts_knit$set(width=75)
```

**Load the required libraries**
```{r}
      Packages <- c("dglm", "hglm", "gdata","readR","tidyr","ggplot",
                    "qqman", "CMplot",  "dplyr", "ggplot2", "SNPRelate",
                    "ggcorrplot", "statmod")
    lapply(Packages, library, character.only = TRUE)

```
**Note: Before showing the codes for DGLM and HGLM analysis, I will first begin with the principal component (PC) analysis of marker data and extract the top fours PCs which is required later for DGLM analysis to account for population structure**

## **Principal component analysis (PCA)**
* In this section we will extract the PC in package SNPRelate using SNP marker data which latter will be used in DGLM analysis to account for population structure.

```{r}
      
# read the vcf file from the folder
    vcf.hww <- "~/Box/Postdoc/Dataanalysis/Wheat_HWW_panel/Data_final/genotype.vcf"
# Reformat the data 
    genofile<-snpgdsVCF2GDS(vcf.hww, "hww.gds", method="biallelic.only")
#Run PCA
    geno_hww<- snpgdsOpen(genofile)
    pca_hww1<- snpgdsPCA(geno_hww, num.thread=2)
# variance proportion (%)
    pc.percent <- pca_hww1$varprop*100
    head(round(pc.percent, 2))
## make a data.frame
    pca_data<- data.frame(sample.id = pca_hww1$sample.id,
                      EV1 = pca_hww1$eigenvect[,1],    # the first eigenvector
                      EV2 = pca_hww1$eigenvect[,2],    # the second eigenvector
                      EV3 = pca_hww1$eigenvect[,3],
                      EV4 = pca_hww1$eigenvect[,4],
                      EV5 = pca_hww1$eigenvect[,5],
                      stringsAsFactors = FALSE)
    head(pca_data)
#save the pca data as r objective for future use
saveRDS(pca_data, file="~/Box/Postdoc/Dataanalysis/Wheat_HWW_panel/results/outputs/
            PCA/pca_hww_299.rds")
# Draw the plot
    plot(pca_data$EV2, pca_data$EV1, xlab="PC1",  col="blue",ylab="PC2",
      yaxt="n",font.lab = 2, col.axis="black", cex.axis=1, tck=-0.02)
```

